File: 2023/day11/part1.py
from itertools import islice


with open("input.txt") as f:
    lines = [[x for x in l] for l in f.read().splitlines()]

z = iter(enumerate(lines))
for idy, line in z:
    if len([x for x in line if x != "."]) == 0:
        lines.insert(idy, line)
        list(islice(z, 2))

print(len(lines))

idx = 0
while idx < len(lines[0]):
    if len([line[idx] for line in lines if line[idx] != "."]) == 0:
        for line in lines:
            line.insert(idx, line[idx])
        idx += 2
    idx += 1

print(len(lines[0]))
        

point = []

for idy in range(len(lines)):
    for idx in range(len(lines[idy])):
        if lines[idy][idx] == "#":
            point.append((idy, idx))

s = 0
for i, a in enumerate(point):
    for b in point[i+1:]:
        s += abs(a[1]-b[1]) + abs(a[0]-b[0])

print(s)

File: 2024/day2/part1.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func check_decrease(input []int, idx int, bad_blocks int) bool {
	if idx+1 >= len(input) {
		return true
	} else if input[idx]-input[idx+1] <= 3 && input[idx]-input[idx+1] > 0 {
		return check_decrease(input, idx+1, bad_blocks)
	} else if bad_blocks == 0 {
		if idx == 0 && check_decrease(input, idx+1, 1) {
			return true
		}
		inputCopy := make([]int, len(input))
		copy(inputCopy, input)
		newList2 := append(input[:idx], input[idx+1:]...)
		newList := append(inputCopy[:idx+1], inputCopy[idx+2:]...)
		fmt.Println(newList)
		fmt.Println(newList2)
		return check_decrease(newList, 0, 1) || check_decrease(newList2, 0, 1)
	}
	return false
}

func check_increase(input []int, idx int, bad_blocks int) bool {
	if idx+1 >= len(input) {
		return true
	} else if input[idx]-input[idx+1] >= -3 && input[idx]-input[idx+1] < 0 {
		return check_increase(input, idx+1, bad_blocks)
	} else if bad_blocks == 0 {
		if idx == 0 && check_increase(input, idx+1, 1) {
			return true
		}
		inputCopy := make([]int, len(input))
		copy(inputCopy, input)
		newList2 := append(input[:idx], input[idx+1:]...)
		newList := append(inputCopy[:idx+1], inputCopy[idx+2:]...)
		fmt.Println(newList)
		fmt.Println(newList2)
		return check_increase(newList, 0, 1) || check_increase(newList2, 0, 1)
	}
	return false
}

func main() {
	// 1. Read input
	input, err := os.ReadFile("input.txt")
	if err != nil {
		log.Fatal(err)
	}
	lines := strings.Split(string(input), "\n")

	safe := 0

	for _, line := range lines {
		linedata := []int{}
		numbers := strings.Split(line, " ")
		for _, number := range numbers {
			conv, _ := strconv.Atoi(number)
			linedata = append(linedata, conv)
		}
		if len(linedata) < 2 {
			continue
		}
		fmt.Println(linedata)
		linedatacopy := make([]int, len(linedata))
		copy(linedatacopy, linedata)
		if check_decrease(linedata, 0, 0) || check_increase(linedatacopy, 0, 0) {
			fmt.Println("safe")
			safe++
		} else {
			fmt.Println("unsafe")
		}
	}

	fmt.Println(safe)
}
